### PR TITLE
Reduce memory usage of itype

### PR DESCRIPTION
**Pull request by ifreund**
#### Summary
SUMMARY: Performance "Reduce memory usage of itype"

#### Purpose of change
The `cata::optional` islots currently stored in `itype` take up their full amount of memory even if they are empty for the item in question.  This results in lots of wasted memory, and the problem will only get worse as we continue to add more items and islots.

#### Describe the solution
Replace `cata::optional` with `std::unique_ptr`.  Dynamically allocating the memory allows us to save memory when slots are unused.

#### Describe alternatives you've considered
This could be left as-is. It's not dysfunctional, but could be improved on.

#### Testing
The game loads successfully and everything seems normal from basic in-game testing. 
The test cases pass for me locally.
The memory savings from rudimentary profiling with `htop` seem to be around 15Mib, which is admittedly not as much as I had hoped for but is better than nothing.

#### Additional context
The main downside to this change is that it requires adding a considerable amount of boilerplate code to `itype` in the form of a copy function.  I'll leave it up to @kevingranade to make the call on whether this is the right move for the codebase or not.

I also had one strange thing happen when loading the game after finishing these changes.  I got the following error on load:
```
 DEBUG    : warnings for type catfood:
invalid stack_size 0 on type using charges


 FUNCTION : void Item_factory::check_definitions() const
 FILE     : src/item_factory.cpp
 LINE     : 1163
```
Upon opening up the json I was surprised to find two separate definitions for the `itype_id` `"catfood"`.  I decided to keep the one [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/34ac4b836c585151b593e38294c7cc3ecc685956/data/json/items/comestibles/other.json#L608) as it seemed to be more fleshed out and in the right place.  Deleting the duplicate entry caused the error to disappear and the game now loads without errors or any noticeable changes from the current master. I'm still not sure exactly why we aren't getting the above debug message on master though, but I thought I'd open this PR up for review while I investigate.

### CODE REVIEW

**Comment by DaviBones** [comment]
Not sure if this helps but I also found and deleted a [duplicate JSON entry for condensed milk](https://github.com/CleverRaven/Cataclysm-DDA/pull/35433/files#diff-98f802a9eeb55475b181e36e02002d85L298) recently. Surely there's some code somewhere that checks for duplicate IDs?

**Comment by BevapDin** [comment]
Would be much easier if you added a `copyable_uniqe_ptr` class and used that instead of the `cata::optional`.  That new class could define the `has_value` function (so you don't need to change code that calls it) and it could implement the copy constructor so you don't need to write an explicit copy constructor for `itype`.

**Comment by kevingranade** [comment]
I was initially against that approach, but had a chat with @jbtw that convinced me I was overreacting to the name copyable_unique_ptr being confusing.

File: src/item.h [hunk 1]
```diff
@@ -1966,7 +1966,7 @@ class item : public visitable<item>
         std::string get_owner_name() const;
         int get_min_str() const;
 
-        const cata::optional<islot_comestible> &get_comestible() const;
```
Comment (kevingranade) [inline src/item.h]: Aaah this is so gross, not in scope for this PR though.

**Comment by kevingranade** [comment]
I'm fine with "copyable_unique_ptr" with a better name, but I do 't want to hold this up, if someone wants to patch that in I'll merge it.

**Comment by kevingranade** [comment]
Clang tidy is complaining about one of the smart pointer invocations.
```
/home/travis/build/CleverRaven/Cataclysm-DDA/src/ranged.cpp:1171:58: error: redundant get() call on smart pointer [readability-redundant-smartptr-get,-warnings-as-errors]

            if( ( pc.last_target.expired() || !pc.sees( *pc.last_target.lock().get() ) ) &&

                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                         pc.last_target.lock()
```

**Comment by ifreund** [comment]
Alright, I made a new pointer type for this purpose, hopefully `cata::value_ptr` is an acceptable name.  I do like not having that massive chunk of boilerplate.  I opted not to add a `has_value()` function so that the new type would have exactly the same semantics as `std::unique_ptr()` and perhaps be a little more familiar to people using it in the future.

File: src/value_ptr.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+template <class T>
+class value_ptr : public std::unique_ptr<T>
+{
+    public:
+        value_ptr() = default;
+        value_ptr( T *value ) : std::unique_ptr<T>( value ) {}
+        value_ptr( const value_ptr<T> &other ) :
+            std::unique_ptr<T>( other ? new T( *other ) : nullptr ) {}
+        value_ptr &operator=( value_ptr<T> other ) {
+            std::unique_ptr<T>::reset( other ? new T( *other ) : nullptr );
```
Comment (BevapDin) [inline src/value_ptr.h]:
```suggestion
            std::unique_ptr<T>::operator=( std::move( other ) );
```
Currently the value is copied twice: once to construct the argument and again within the operator.

Comment (ifreund) [reply src/value_ptr.h]: Thanks, that's certainly cleaner

File: src/value_ptr.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * This class is essentially a copyable unique pointer. Its purpose is to allow
+ * for sparse storage of data without wasting memory in classes such as itype.
+ */
+template <class T>
+class value_ptr : public std::unique_ptr<T>
+{
+    public:
+        value_ptr() = default;
+        value_ptr( T *value ) : std::unique_ptr<T>( value ) {}
+        value_ptr( const value_ptr<T> &other ) :
```
Comment (jbytheway) [inline src/value_ptr.h]: I think you need an explicitly-defaulted move constructor too.  If you want to double-check, you could add a unit test that a moved-from `value_ptr` is empty (which would be a good thing to have in any case).

Comment (ifreund) [reply src/value_ptr.h]: you were totally right @jbytheway, I defaulted the move constructor and added tests for all of the functions.

File: tests/value_ptr.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    CHECK( !a );
+    CHECK( !!b );
+}
+
+TEST_CASE( "value_ptr move constructor", "[value_ptr]" )
+{
+    cata::value_ptr<int> a = cata::make_value<int>( 7 );
+    REQUIRE( !!a );
+    cata::value_ptr<int> b( std::move( a ) );
+    CHECK( !a );
```
Comment (jbytheway) [inline tests/value_ptr.cpp]: You will probably need to add a suppression here to avoid a clang-tidy warning about use-after-move.

Comment (ifreund) [reply tests/value_ptr.cpp]: Alright, I believe I've suppressed it correctly

**Comment by ifreund** [comment]
Alright, looks like this passed the clang-tidy checks on travis.  Should be good to go now.

**Comment by ifreund** [comment]
Conflicts resolved.

**Comment by ifreund** [comment]
More conflicts resolved, got unlucky with the timing on this one I guess
